extensions-core/s3-extensions/src/main/java/org/apache/druid/storage/s3/output/S3StorageConnectorModule.java

```diff
@@ -0,0 +1,44 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.storage.s3.output;
+
+import com.fasterxml.jackson.databind.Module;
+import com.fasterxml.jackson.databind.module.SimpleModule;
+import com.google.inject.Binder;
+import org.apache.druid.initialization.DruidModule;
+
+import java.util.Collections;
+import java.util.List;
+
+public class S3StorageConnectorModule implements DruidModule
+{
+  @Override
+  public List<? extends Module> getJacksonModules()
+  {
+    return Collections.singletonList(
+        new SimpleModule(this.getClass().getSimpleName()).registerSubtypes(S3StorageConnectorProvider.class)
+    );
+  }
+
+  @Override
+  public void configure(Binder binder)
+  {
+  }
```

Comment:
Trying to sort out how this works. I guess I'd expect the "connector" to be a singleton in Guice. That connector can have items injected into it, including its config. It might have a lifecycle, so it can delete files on start and exit. It might have a secrets provider, such as for S3. Then each request can open a file, say, based on that connector, using maybe a "connection" or a "session". That session would also delete its files on session close. The session can give rise to streams or chunked operations.

Something like that is the usual way these things work. I'm wondering if the JSON configuration gives us the flexibility these kinds of things usually need.

Reply:
I haven't tried it but I think `PolyBind` contains methods for doing this. Every storage connector can bind to a `PolyBind.optionalBinder` with the scheme it supports.
And the consumer can then pull the the storage connector it needs via `PolyBind.createChoice` where the consumer property which generate the choice.
An example of this happens with MetricsFactory injection, if that's analogous to this usecase.